Add support for specifying compression level

dmitryvk opened the pull request:
Hi,

I've been using zip-rs to compress somewhat large files on the fly.
Setting compression level from 6 (the default) to 1 makes a big difference in time to compress (about ~3-fold reduction).

In #212 some concerns regarding API compatibility were raised, but I don't see any concerns: field is added to ZipFileData which is not exported from the crate.

I also don't like using u32 to specify compression level (without range checking) - but it seems that flate2, bzip and zstd also use u32.

zamazan4ik commented on src/write.rs:
```diff
@@ -143,6 +151,20 @@ impl FileOptions {
+    /// * `Bzip2`: 0-9. `None` corresponds to 6
+    /// * `Zstd`: 1-21. `None` corresponds to 3
+    /// * others: only `None` is allowed
+    /// `None` value specifies default compression level.
```
I suggest to leave the information about compression level only in this place and remove duplication in other places. It will reduce complication in future maintenance.

dmitryvk replied:
Agreed; removed other doc comments

zamazan4ik commented:
Thanks for the PR! I am really glad to see this kind of functionality in the library.

Regarding range check question. Honestly, I also would like to see some kind of range checking. Maybe we can at least provide to the user some kind of helper. But in this case, we will introduce some additional kind of coupling between our library and dependent crates (we will need our range checkers with their supported ranges for each algorithm). I cannot say right now, which way is better.

dmitryvk commented:
I've updated the PR.
Extra doc comments are removed.
I've added range checking for compression level. Values for deflate and bzip2 seem to be standardized, but zstd is somewhat complicated.
It seems that over time compression level range of zstd was increased: zstd crate mentions levels 1-21 and zstd docs state that maximum is 22 and negative values are allowed. Wikipedia states that -7 is the lowest value. Looking at zstd source code it seems that the actual limit is much lower (seems to be -128K though I didn't 100% verify it). Relaxing the constraints is easier so I went with more conservative range of -7 to 22.

dmitryvk commented:
> But in this case, we will introduce some additional kind of coupling between our library and dependent crates

Yes, this is a coupling, but it's already present. Also, all of the crates provide means to query for allowed compression level values (`flate2::Compression::none/best`, `bzip2::Compression::none/best`, `zstd::compression_level_range`)

zamazan4ik commented:
> Yes, this is a coupling, but it's already present. Also, all of the crates provide means to query for allowed compression level values (flate2::Compression::none/best, bzip2::Compression::none/best, zstd::compression_level_range)

Yeah, that is what I wanted to see - using for the range checks ranges directly from our dependencies and does not introduce such ranges in our codebase. 

LGTM.

Since this contribution is quite important from the feature perspective, I suggest wait for some time for the review from @Plecra

Plecra reviewed (approved):
I like it :) Thanks for the changes

dmitryvk commented:
Thanks, I didn't expect such quick response:)
